### PR TITLE
avoid deadlock when sql engine if full with coordinators

### PR DESCRIPTION
**Pull request by dorinhogea**
Bug: if all sql engines in a node are running coordinators for parallel sql execution, there is no progress since workers have no sql engines to run on.  Therefore, while the coordinators are blocked in sql engines waiting for workers to proceed, workers are waiting in queue for additional sql engines.  
Fix: we introduce a new configurable setting, "**dohsql_pool_thread_slack**", that reserves a number of sql engines for non-coordinator work.  It is never less than 1.  
In a nutshell, the reserved engines are able to drain the sql queue, which includes parallel sql workers, and allow the dohsql coordinators to make progress.

Signed-off-by: Dorin Hogea <dhogea@bloomberg.net>

### CODE REVIEW

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
@@ sqlite/ext/comdb2/tables.c @@
+ If a new system table is added, be sure to document it in docs/pages/programming/system_tables.md.
@@ sqlite/src/parse.y @@
+ Custom tokens and rules must be defined after SQLite's.
+ Please update SQL documentation and bubble diagrams to reflect the changes you've made, if necessary.
@@ sqlite/src/vdbeInt.h @@
+ If the order of members in struct Mem is changed, columnNullValue (defined in sqlite/vdbeapi.c) needs to be reordered accordingly.
@@ sqlite/src/vdbeaux.c @@
+ If a new fixed-length serial type is introduced, sqlite3IsFixedLengthSerialType (defined in sqlite/vdbeInt.h) needs to be changed accordingly.
@@ db/prefault_helper.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ db/sqlglue.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ db/appsock_handler.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ db/fdb_bend.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ db/handle_buf.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ schemachange/sc_global.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ util/walkback.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ util/logmsg.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/0652541e/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>3/343 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>socksql_master_swings [core dumped]<br>odh_blobs<br>tunables</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/0652541e/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/0652541e/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2841'>Resubmit</a></p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
@@ sqlite/ext/comdb2/tables.c @@
+ If a new system table is added, be sure to document it in docs/pages/programming/system_tables.md.
@@ sqlite/src/parse.y @@
+ Custom tokens and rules must be defined after SQLite's.
+ Please update SQL documentation and bubble diagrams to reflect the changes you've made, if necessary.
@@ sqlite/src/vdbeInt.h @@
+ If the order of members in struct Mem is changed, columnNullValue (defined in sqlite/vdbeapi.c) needs to be reordered accordingly.
@@ sqlite/src/vdbeaux.c @@
+ If a new fixed-length serial type is introduced, sqlite3IsFixedLengthSerialType (defined in sqlite/vdbeInt.h) needs to be changed accordingly.
@@ db/prefault_helper.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ db/sqlglue.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ db/appsock_handler.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ db/fdb_bend.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ db/handle_buf.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ schemachange/sc_global.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ util/walkback.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ util/logmsg.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/7f6ad7d1/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: Success &#10003;.</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/7f6ad7d1/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/7f6ad7d1/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2841'>Resubmit</a></p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
@@ sqlite/ext/comdb2/tables.c @@
+ If a new system table is added, be sure to document it in docs/pages/programming/system_tables.md.
@@ sqlite/src/parse.y @@
+ Custom tokens and rules must be defined after SQLite's.
+ Please update SQL documentation and bubble diagrams to reflect the changes you've made, if necessary.
@@ sqlite/src/vdbeInt.h @@
+ If the order of members in struct Mem is changed, columnNullValue (defined in sqlite/vdbeapi.c) needs to be reordered accordingly.
@@ sqlite/src/vdbeaux.c @@
+ If a new fixed-length serial type is introduced, sqlite3IsFixedLengthSerialType (defined in sqlite/vdbeInt.h) needs to be changed accordingly.
@@ db/prefault_helper.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ db/sqlglue.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ db/appsock_handler.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ db/fdb_bend.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ db/handle_buf.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ schemachange/sc_global.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ util/walkback.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ util/logmsg.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/cb7e2a85/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: Success &#10003;.</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/cb7e2a85/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/cb7e2a85/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2841'>Resubmit</a></p>

**Review by adizaimi — APPROVED** [review]
LGTM!

**Comment by dorinhogea** [comment]
TYA
